### PR TITLE
opt: calculate stats for partial spatial indexes

### PR DESCRIPTION
**Pull request by mgartner**
This commit updates the statistics builder so that partial index
predicates of spatial indexes are considered when calculating the
statistics for spatial index scans.

Release note: None

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MHqH8mV0g25Il4S1Piz:b-7cmhvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
> ```
>       │              │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x01", "B\xfd\x12\x00\x00\x00\x00\x00\x00\x00")
>       │              │         └── ["B\xfd\x14\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x14\x00\x00\x00\x00\x00\x00\x00"]
>       │              ├── stats: [rows=556.5, distinct(1)=159, null(1)=0, distinct(3)=3, null(3)=0, distinct(6)=3, null(6)=0, distinct(3,6)=9, null(3,6)=0]
> ```

Why are there more rows in this case than for the tests without nulls? Shouldn't this be the same?
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1398 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqHKc-3j4VWrfjeDIL:-MHqHKc-3j4VWrfjeDIM:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1398)):*
> ```
>       │              │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x01", "B\xfd\x12\x00\x00\x00\x00\x00\x00\x00")
>       │              │         └── ["B\xfd\x14\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x14\x00\x00\x00\x00\x00\x00\x00"]
>       │              ├── stats: [rows=556.5, distinct(1)=159, null(1)=0, distinct(3)=3, null(3)=0, distinct(6)=3, null(6)=0, distinct(3,6)=9, null(3,6)=0]
> ```

ditto

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MHr9wxa9FBx_Btvsc6I:bpunqjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why are there more rows in this case than for the tests without nulls? Shouldn't this be the same?
</blockquote></details>

It seems to be the same issue as this: https://github.com/cockroachdb/cockroach/issues/51396. I suppose now is the time to fix it!

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MI1WUKRAs4pfuM4Ab6F:b-d3mmjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It seems to be the same issue as this: https://github.com/cockroachdb/cockroach/issues/51396. I suppose now is the time to fix it!
</blockquote></details>

I'm still trying to get to the bottom of this. I think I was wrong in my last comment—that is _an_ issue but fixing it didn't fix this discrepancy. 

The problem appears to be that in the case where the table stats have `0` nulls for column `s`, `sb.selectivityFromMultiColDistinctCounts(constrainedCols, scan, s)` and `sb.selectivityFromSingleColDistinctCounts(histCols, scan, s)` return the same values. Because we apply the inverse of the latter result, they cancel each other out.

But when the table has a non-zero number of nulls for column `s`, they no longer return the same result, so after applying the first and the inverse of the second, the result row count diverges from the no-nulls case.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MI1Z-al2kgqRuxESV9M:bfbap10) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm still trying to get to the bottom of this. I think I was wrong in my last comment—that is _an_ issue but fixing it didn't fix this discrepancy. 

The problem appears to be that in the case where the table stats have `0` nulls for column `s`, `sb.selectivityFromMultiColDistinctCounts(constrainedCols, scan, s)` and `sb.selectivityFromSingleColDistinctCounts(histCols, scan, s)` return the same values. Because we apply the inverse of the latter result, they cancel each other out.

But when the table has a non-zero number of nulls for column `s`, they no longer return the same result, so after applying the first and the inverse of the second, the result row count diverges from the no-nulls case.
</blockquote></details>

In the no-nulls case, the `singleColSelectivity` and `multiColSelectivity` have the same values within `selectivityFromMultiColDistinctCounts`, so the weighted return value it equal to their shared value.

In the nulls case, the `singleColSelectivity` is half the value of `multiColSelectivity` meaning the weighted return value could never be equal to the return value of `selectivityFromSingleColDistinctCounts`.

Another puzzle to me is why this doesn't affect JSON and ARRAY partial inverted index scans and regular partial index scans. The implementation for them is basically the same, so it's unclear to me why the results are different.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MI1zDiYCS__HjGg0Ch2:b-97qvoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

In the no-nulls case, the `singleColSelectivity` and `multiColSelectivity` have the same values within `selectivityFromMultiColDistinctCounts`, so the weighted return value it equal to their shared value.

In the nulls case, the `singleColSelectivity` is half the value of `multiColSelectivity` meaning the weighted return value could never be equal to the return value of `selectivityFromSingleColDistinctCounts`.

Another puzzle to me is why this doesn't affect JSON and ARRAY partial inverted index scans and regular partial index scans. The implementation for them is basically the same, so it's unclear to me why the results are different.
</blockquote></details>

It looks like it _does_ affect other partial index scans. It even affects Select stats. So it doesn't seem specific to partial or spatial or partial-spatial indexes. Examples: https://gist.github.com/mgartner/d2911f90c99c13c6c4ef91e7585c0668 (I think the values in these examples go down because of a bug with distinct count, but I believe [backed by a rough attempt at fixing the bug and some testing] that they would go up, similar to the test cases in this PR, if the bug is fixed).

Is it possible that this is a limitation of `selectivityFromMultiColDistinctCounts`? Or maybe a bug, but it's hard for me to tell—it seems like what it's doing is sane.

Or maybe its a limitation/bug of `predicateSelectivity`: https://github.com/cockroachdb/cockroach/blob/66a076f89e92becc1d7e1a2ecfd39c275165e2d8/pkg/sql/opt/memo/statistics_builder.go#L3747. When `nullSelectivity` is zero, as `inputNullCount` increases, `outputRows`  and `sel` decrease. This makes sense when there are no histograms, but it doesn't make as much sense when there are—if a histogram tells you that 100 rows have a value of 'banana` for the column, the number of null values for that column in the other rows shouldn't affect the selectivity. At least, I think... It's getting far too late to reason about statistics. :upside_down_face:

**Review by rytaft — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-:-MIKRfmgEJyJn3hbcZPX:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHqH8mV0g25Il4S1Piy:-MIKR9whAD9u2YFsXW7M:bcnymxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/98618411eda780a637bc4dc73ebbd894bb909c9e/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1348)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It looks like it _does_ affect other partial index scans. It even affects Select stats. So it doesn't seem specific to partial or spatial or partial-spatial indexes. Examples: https://gist.github.com/mgartner/d2911f90c99c13c6c4ef91e7585c0668 (I think the values in these examples go down because of a bug with distinct count, but I believe [backed by a rough attempt at fixing the bug and some testing] that they would go up, similar to the test cases in this PR, if the bug is fixed).

Is it possible that this is a limitation of `selectivityFromMultiColDistinctCounts`? Or maybe a bug, but it's hard for me to tell—it seems like what it's doing is sane.

Or maybe its a limitation/bug of `predicateSelectivity`: https://github.com/cockroachdb/cockroach/blob/66a076f89e92becc1d7e1a2ecfd39c275165e2d8/pkg/sql/opt/memo/statistics_builder.go#L3747. When `nullSelectivity` is zero, as `inputNullCount` increases, `outputRows`  and `sel` decrease. This makes sense when there are no histograms, but it doesn't make as much sense when there are—if a histogram tells you that 100 rows have a value of 'banana` for the column, the number of null values for that column in the other rows shouldn't affect the selectivity. At least, I think... It's getting far too late to reason about statistics. :upside_down_face: 
</blockquote></details>

Yep, I see the problem. Thanks for digging into this! I'll try to fix this issue separately. I think this PR is good to go.

File: pkg/sql/opt/memo/statistics_builder.go [hunk 1]
```diff
@@ -831,14 +832,41 @@ func (sb *statisticsBuilder) invertedConstrainScan(scan *ScanExpr, relProps *pro
...
+	// determine not-null columns from the constraint. However, the partial
+	// index predicate may guarantee that non-indexed columns are not-null.
+	notNullCols := relProps.NotNullCols.Copy()
+	if pred != nil {
+		// Add any not-null columns from the predicate constraints.
+		for i := range pred {
+			if c := pred[i].ScalarProps().Constraints; c != nil {
+				cols := c.ExtractNotNullCols(sb.evalCtx)
+				const distinctCount = math.MaxFloat64
+				sb.ensureColStat(cols, distinctCount, scan, s)
```
Comment (mgartner) [inline pkg/sql/opt/memo/statistics_builder.go:847]: @rytaft Just to be clear, you're ok with calling `ensureColStat` here, before `sb.applyFilter` (which eventually leads to `updateDistinctCountFromHistogram` being called). This prevents incorrect distinct counts.

**Review by rytaft — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 847 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MHm4Oy6-x-1GquRIDA2-r2-847:-MIzHvyZ1Zy6LJcVTRVY:b-2nqbll) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1315e75096bd572379900315a3c7b1f84ff013b/pkg/sql/opt/memo/statistics_builder.go#L847)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

@rytaft Just to be clear, you're ok with calling `ensureColStat` here, before `sb.applyFilter` (which eventually leads to `updateDistinctCountFromHistogram` being called)? This prevents incorrect distinct counts. 
</blockquote></details>

Seems reasonable to me....
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MIzLVTaBWK4ftUHp-jW:-MIzLVTaBWK4ftUHp-jX:b-fsh5z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b6b24c84b5317266f53782a0cdb480bb86fac4f/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1547)):*
> ```
>       ├── key: (1)
>       ├── fd: (1)-->(2,3)
>       ├── scan spatial
> ```

How come we don't prefer the partial index here?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MIzLVTaBWK4ftUHp-jW:-MIzenIaCoPVErOsXs8s:bqx97ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b6b24c84b5317266f53782a0cdb480bb86fac4f/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1547)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How come we don't prefer the partial index here?
</blockquote></details>

It looks like the high number of nulls was making the estimated number of rows scanned higher than it should be, due to the issue with single- and multi-column distinct count selectivity we spoke about. I've lowered the number of nulls to `100` and now the partial index scan is selected.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-:-MIztnQoBhY5MG5fhXHa:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 1547 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54627#-MIzLVTaBWK4ftUHp-jW:-MIztSI3EeIyFME0MOgV:b-l2keul) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b6b24c84b5317266f53782a0cdb480bb86fac4f/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L1547)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It looks like the high number of nulls was making the estimated number of rows scanned higher than it should be, due to the issue with single- and multi-column distinct count selectivity we spoke about. I've lowered the number of nulls to `100` and now the partial index scan is selected.
</blockquote></details>

Cool - thanks!

**Comment by mgartner** [comment]
bors r=rytaft

**Pull request merged**